### ISSUE TITLE
Japonicus on Windows

### ISSUE THREAD

**Issue by jthope** [Gab0/japonicus#97]
Japonicus will install on windows, however it will not run due to parallel processing and how the japonicus.py file is setup (more info: https://stackoverflow.com/questions/18204782/runtimeerror-on-windows-trying-python-multiprocessing).

I modified the file a bit to protect the main code from recursive sub-process calls. I then ran into other issues with logging (probably file system related).

My question - is there any interest in supporting windows? And if there is, would you take a pull request if I managed to figure it out? Right now I am running a Linux Mint VM, but would be awesome if I could backtest natively on my main OS.

**Comment by Gab0** [comment]
Once there was an issue opened about this. I can't test on windows and I think the win userbase is low (it actually does not work on windows, right?). If you can fix this stuff with elegance, the pull request will be accepted of course.

**Comment by jthope** [comment]
I was able to get japonicus.py straightened out by doing the following:
![image](https://user-images.githubusercontent.com/114626/37103770-fe502b90-21f8-11e8-95cf-0cc221a3a1d4.png)

However, I'm still fighting with the evolution_generations.py file. It freezes the console. It's not clear how I can suppress the subprocess from executing on the main module (according to above stackoverflow link).

I was able to get the file pathing to work by doing the following on the HOME variable. However, I am not versed in python (my focus is in C#, javascript) so I am sure this can be done a better way:
HOME = os.path.abspath(os.path.join(os.path.dirname(os.path.abspath(__file__)), os.pardir))

I have to get my IDE setup for python development in windows, and will come back to this. Getting a solid strat running is my first priority.

**Comment by Gab0** [comment]
Well, wrapping the entry of japonicus in a main function is actually a very good idea, but seems you wrapped only the start of the procedures. Do you think this may help on windows compatibility? Dunno, anyway I'm doing this cleanse on the next commit; This method file pathing method will also be online.
From what I read we would need to make major changes on the subprocess methods to handle windows.. maybe at first just make a fallback to non-parallel methods. plz do a small research on the subject ^^
cheers!

**Comment by Gab0** [comment]
This HOME line is to replace `chdir @ japonicus.py:19` or to replace that HOME on `Settings.py`?
I tested this, it returns `/home` on Linux which would not be good on either case... please elaborate this subject.

**Comment by jthope** [comment]
Re: the HOME variable, that was in settings.py -- technically I can set this path by setting a home variable in windows, but it would be easier to just clone/compile/run. That line worked for windows but I have yet to test it on linux as I wasn't very happy with the syntax to begin with. There is probably a better way to get the parent directory path from the executing application path. I will look into this closer for a solution that works cross platform. 

In regards to the subprocess methods, from my understanding, in order for python multiprocessing to work in windows, the main code needs to be protected to avoid recursive subprocess creation. Looking at [https://docs.python.org/3/library/multiprocessing.html](url) shows multiprocessing being done a bit differently than in japonicus. Reading the issues on stackoverflow have a lot of python developers saying it's an issue with python, and the other half are saying it's an issue with the code implementation. But I don't have enough knowledge around python to make that assumption.

If I don't run japonicus in parallel mode, then the issues come down to file pathing. For a quick test I hardcoded the paths, and japonicus ran fine. So I think the only thing stopping this from running on windows is the way parallel processing was implemented and how the file paths are set.

I'll play more with it this weekend to see if I can make some progress/provide better input. I check for updates daily so my local clone will be using the latest code.

**Comment by Gab0** [comment]
Hey, the home thing can apparenty be solved with a new lib on Python3.5+:
```
from pathlib import Path
home = str(Path.home())
# https://stackoverflow.com/questions/4028904/how-to-get-the-home-directory-in-python
```
I can't test it but I trust these guys of stackoverflow, this snippet will be featured on next update. 
The home detection is actually not very important I think, most users will have to manually set their own paths..
The subprocess problem is very serious to blindly solve, so I will at least simplify the code (currently a maze) for you to fix if u can . If needed I find some windows to test it..
cheers!

**Comment by nmikaty** [comment]
Hi, did you find a solution for parallel processing on Windows?
I get this error and suspect this is what you are talking about:

`Process SpawnPoolWorker-1:
Traceback (most recent call last):
  File "C:\Users\xxxxxx\AppData\Local\Programs\Python\Python36\lib\multiprocessing\process.py", line 258, in _bootstrap
    self.run()
  File "C:\Users\xxxxxx\AppData\Local\Programs\Python\Python36\lib\multiprocessing\process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\xxxxxx\AppData\Local\Programs\Python\Python36\lib\multiprocessing\pool.py", line 108, in worker
    task = get()
  File "C:\Users\xxxxxx\AppData\Local\Programs\Python\Python36\lib\multiprocessing\queues.py", line 337, in get
    return _ForkingPickler.loads(res)
AttributeError: Can't get attribute 'Individual' on <module 'deap.creator' from 'C:\\Users\\xxxxxx\\AppData\\Local\\Programs\\Python\\Python36\\lib\\site-packages\\deap\\creator.py'>`

I saw that japonicus.py has been cleaned up already and the `home = str(Path.home())` implemented. 

I have another error, but I think it's related to the dash web server:

`C:\Users\xxxxxx\AppData\Local\Programs\Python\Python36\lib\site-packages\plotly\tools.py:103: UserWarning:

Looks like you don't have 'read-write' permission to your 'home' ('~') directory or to our '~/.plotly' directory. That means plotly's python api can't setup local configuration files. No problem though! You'll just have to sign-in using 'plotly.plotly.sign_in()'. For help with that: 'help(plotly.plotly.sign_in)'.
Questions? Visit https://support.plot.ly`

Thank you for this great piece of code and your help!
Hope I can get it working on Windows so I don't have to setup another server on AWS :)

**Comment by Gab0** [comment]
Hi @nmikaty, yeah this traceback is related to Windows.
Last time I searched the docs and found info on the windows issue, can't find it anymore. If you are able to point it to me, I can check it and update the development branch for you to test. 
The second one is about plotly on windows, and seems like a non important warning.
Thanks.

**Comment by nmikaty** [comment]
I think this answers most questions: https://github.com/DEAP/deap/issues/57
But I'm not sure how to implement it with your code. Also, it's my first time with Python so I'm a bit lost :/

**Comment by Gab0** [comment]
Thats really the source. Its a problem with deap.creator...
They made a fork to fix the error on DEAP, but that fork became old (3y). The issue there about this problem has recent activity, so they may be fixing it soon and solve our problems.

The bugsolver mod consists of this:
https://github.com/cmd-ntrf/deap-1/commit/c71b773d442007719ba530a140f6159789a14058

I will try to apply the patch on japonicus itself to guarantee, altough I have no idea of the best way to do this. If someone could apply it, would be good. Cheers!

**Comment by Gab0** [comment]
@nmikaty Hey, this new error is another on py2->py3 changes. `.iteritems` is now `.items`, can't you try to change that? If you deleted the patch and fully rolled back its ok, leave it alone ^^.
Cheers!

**Comment by Gab0** [comment]
@nmikaty Yeah just `.items` was to be changed. 
The fix is now on japonicus itself, no need to mess with DEAP.
I think DEAP team is releasing a new version soon but it doesn't matter.
 @nmikaty @jthope please confirm if it runs natively under Windows..

**Comment by nmikaty** [comment]
Yes it works! (haven't updated your code yet, but the change `iteritem` > `item` did it)

I still have this annoying warning all over my screen, but the rest seems to be working fine:
```
C:\Users\xxxxxx\AppData\Local\Programs\Python\Python36\lib\site-packages\plotly\tools.py:103: UserWarning:

Looks like you don't have 'read-write' permission to your 'home' ('~') directory or to our '~/.plotly' directory. That means plotly's python api can't setup local configuration files. No problem though! You'll just have to sign-in using 'plotly.plotly.sign_in()'. For help with that: 'help(plotly.plotly.sign_in)'.
Questions? Visit https://support.plot.ly
```

I will try to solve this later and update here.

Thanks and well done @Gab0 !

**Comment by Gab0** [comment]
Nice! So it should work for any windows user that pulls the develop branch of japn, as this entire fixed `creator`class is now incorporated to it.. works on stock DEAP.

**Comment by PrestaNinja** [comment]
Some feedback on the installation process on Windows 10, Python 3.7.2:
Notice: I am not very keen on Python. Probably there is a much proper way to do this. I just wanted to share my experience for future reference.
Given that you already have a C++ compiler installed: 
- The provided installation stack using requirements.txt file won't compile. Instead, probably not the best but the practical method is just to run japonicus and wait for the module dependency errors and then install the missing packages one by one manually.

- The latest versions of dash won't work as the name "Event" was removed by v.0.36.1, nor will 0.21.1. 
I could only get it running with: 
pip3 install dash==0.24.1

**Comment by Gab0** [comment]
Hey bro @PrestaNinja, thanks for the feedback.

Some modules fail to install while inside a requirements.txt because this pipenv (or your global python) does not have numpy installed. This is getting more common, very annoying. The failure was due to missing numpy? On my test it was... Solution is to install `numpy` then the `requirements.txt` batch. I'm trying to find a better way out for this.

`dash==0.24.1` will be at `requirements.txt`.. Thanks.

**Comment by AbinGibiAbin** [comment]
Hi again,

I actually had numpy installed. As far as I can see, it's dash dependencies causing the issue as I needed to downgrade dash to 0.24.1. Otherwise, with newest dash (0.37.1) it would complain about "Event" name. I really have no idea about the details, however I guess the stable solution would be to replace this "Event" stuff with whatever they introduced in place of it, so it would compile without the need to downgrade anything.

**Comment by Gab0** [comment]
Yeah, the correct way to solve is to replace with the newer thing. I thought that it would require some reengineering of the thing, so I postponed it... turns out the Event thing was removed from dash, and was also removed from our web interface long time ago... had just the variable lying around xD so back to latest dash.

Oh, so the requirements.txt thing may be unrelated to numpy? if you could recreate this error and paste the message here it would be great..
Thanks.

**Comment by AbinGibiAbin** [comment]
Here you go, the output is attached. I also tried changing the requirements to dash=0.24.1 but the result is the same. I think this might be strongly related to my own configuration yet I have compiled python wheels on this box before. I hope this helps.
[log.txt](https://github.com/Gab0/japonicus/files/3001943/log.txt)
